device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/auth/IotHubSasTokenProvidedAuthenticationProvider.java

```diff
@@ -0,0 +1,76 @@
+/*
+*  Copyright (c) Microsoft. All rights reserved.
+*  Licensed under the MIT license. See LICENSE file in the project root for full license information.
+*/
+
+package com.microsoft.azure.sdk.iot.device.auth;
+
+import com.microsoft.azure.sdk.iot.device.ClientOptions;
+import com.microsoft.azure.sdk.iot.device.IotHubClientProtocol;
+import com.microsoft.azure.sdk.iot.device.SasTokenProvider;
+
+import javax.net.ssl.SSLContext;
+
+/**
+ * {@link IotHubSasTokenAuthenticationProvider} implementation where the tokens are provided by an instance of {@link SasTokenProvider}.
+ * This is used in cases like when the user creates a device client with {@link com.microsoft.azure.sdk.iot.device.DeviceClient#DeviceClient(String, String, SasTokenProvider, IotHubClientProtocol, ClientOptions)}
+ */
+public class IotHubSasTokenProvidedAuthenticationProvider extends IotHubSasTokenAuthenticationProvider
+{
+    SasTokenProvider sasTokenProvider;
+    char[] lastSasToken;
+
+    public IotHubSasTokenProvidedAuthenticationProvider(String hostName, String deviceId, String moduleId, SasTokenProvider sasTokenProvider, SSLContext sslContext) {
+        super(hostName, null, deviceId, moduleId, sslContext);
+
+        if (sasTokenProvider == null)
+        {
+            throw new IllegalArgumentException("SAS token provider cannot be null");
+        }
+
+        this.sasTokenProvider = sasTokenProvider;
+    }
+
+    @Override
+    public boolean isAuthenticationProviderRenewalNecessary()
+    {
+        // Renewal of the authentication provider itself is never needed since the SAS token provider is responsible
+        // for providing SAS tokens indefinitely.
+        return false;
+    }
+
+    @Override
+    public void setTokenValidSecs(long tokenValidSecs)
+    {
+        throw new UnsupportedOperationException("Cannot configure SAS token time to live when custom SAS token provider is in use");
+    }
+
+    @Override
+    public boolean canRefreshToken()
```

Comment:
As this is overridden and specified for this impl, I think it warrants a comment as to why it is `true`.